### DISCUSSION TITLE
How does ATC handle container/volume worker metrics + rebalancing?

### DISCUSSION THREAD

**Discussion by pn-santos** [concourse/concourse#5992] [Help & Support]
Hi,

We are running a cluster with 3 ATCs and have run into a slight issue with container/volume count that is exposed via prometheus metrics and would be interested in understanding how concourse is _supposed_ to manage these.

Here's a picture of the issue:

![workers](https://user-images.githubusercontent.com/23340138/90388788-49fd6380-e080-11ea-937c-00e5cd9c80ce.png)

The graph shows the container count for a single worker across 3 ATCs (each coloured line is an ATC instance). A couple of things seem obvious:

- Only one ATC reports correct container count at any one time (the one the worker is registered with)
- The last container count an ATC received is never cleared (at least it persists for several hours)

Is this _supposed_ to happen? I had a very brief look [at the code](https://github.com/concourse/concourse/blob/6ab1954b70a7dfd6e27293f0ec6c7f7bb86bb5d7/atc/metric/emitter/prometheus.go#L753-L775) and it _seems_ (I didn't dig too much into it) that if a worker heartbeat is absent for 5 min the ATC should reset all the worker related metrics (which is what would make the most sense) but this doesn't seem to be happening.

Weirdly this seems to happen with `linux` workers but not with `darwin` ones! 

For now we've resorted to some promql magic (thanks @alexandergurney) to try and remove the "flat line" measurements but it's a bit of a hack.

Could this be a bug? Could it be related to some aspect of the cluster setup (configs and whatnot?)

Any insight would be appreciated!

**Comment by pn-santos** [comment]
no ideas?

**Comment by robwhitby** [comment]
I've just stumbled across the same problem after noticing the metrics don't match the output of `fly workers`.

the stale metrics seem to be emitted for varying lengths of time before being cleaned up:

```
concourse_workers_containers{worker="ecaa9a55-05f5-4cc2-9852-56ce0188e670"}
```

<img width="1188" alt="image" src="https://user-images.githubusercontent.com/176209/118520883-59562e00-b732-11eb-93a9-056cf95af876.png">

**Comment by databus23** [comment]
This should be fixed by https://github.com/concourse/concourse/pull/7965 which will be included in the 7.7.0 release.
